### DISCUSSION TITLE
Why not keeping tensors in GPU memory in your diarization model?

### DISCUSSION THREAD

**Discussion by Thytu** [NVIDIA-NeMo/NeMo#7795] [Q&A]
First of all, congrats for your work! We are observing better results than with PyAnnote for 10x less time 🚀 

When analysing the inner working of your diarization model I figured that [during the generation of embeddings](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/models/clustering_diarizer.py#L350-L388) you're actually moving the tensors [back to CPU](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/models/clustering_diarizer.py#L362) and then [saving it as pickle](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/models/clustering_diarizer.py#L388).

Moving this data from GPU to CPU and then saving it as pickle is time consuming and using CPU almost fully the CPU.\
(cf my following screenshot)

Why won't you instead keep the tensors in GPU VRAM and provide it to the next model?\
This would deeply accelerate the overall pipeline and reduce the CPU footprint.

#### <ins>Screenshot of the CPU usage when generating embedding</ins>
![image](https://github.com/NVIDIA/NeMo/assets/43698357/e27d8990-59b1-415a-83c8-dc5e44427a56)

CC @mariano-at-gladia

**Comment by nithinraok** [comment]
Wonderful to hear this `We are observing better results than with PyAnnote for 10x less time` . 
@Thytu thanks for the suggestion, currently you can skip saving embeddings by providing `save_embeddings=False` through config. 

However, we decided to move to CPU as we perform batch processing on datasets and keeping them on GPU would run VRAM out. It might make sense if we run inference on single audio less than 5 mins. Have you tried diarization through RIVA?

**Reply by Thytu** [reply]
Thanks for the reply @nithinraok 

Unfortunately RIVA isn't an options for me as RIVA isn't OSS and I need to be able to tweak the models I use.
Thanks for the `save_embeddings` param, indeed it saves me 1s 🙂 

Is NeMo inference oriented or is it more a general tool for AI?
This batching prevents user to use NeMo diarization pipeline for inference (or are least make it slower that it would needs to be).

Can we imagine an inference mode where we would keep the tensor in GPU?

**Reply by Thytu** [reply]
Update : Diarization is indeed quickly filling the RAM. Instead of going full GPU I tried the following changes:

```diff

+         move_to_cpu_every = 400
+         all_embs = torch.empty([0])
+         in_gpumemory_embs = torch.empty([0], device=self._speaker_model.device)

-       all_embs = torch.empty([0])

        for test_batch in tqdm(
            self._speaker_model.test_dataloader(),
            desc=f'[{scale_idx+1}/{num_scales}] extract embeddings',
            leave=True,
            disable=not self.verbose,
        ):

            test_batch = [x.to(self._speaker_model.device) for x in test_batch]
            audio_signal, audio_signal_len, labels, slices = test_batch

            with autocast():
                _, embs = self._speaker_model.forward(input_signal=audio_signal, input_signal_length=audio_signal_len)
                emb_shape = embs.shape[-1]
                embs = embs.view(-1, emb_shape)

-               all_embs = torch.cat((all_embs, embs.cpu().detach()), dim=0)

+               in_gpumemory_embs = torch.cat((in_gpumemory_embs, embs), dim=0)

+               if len(in_gpumemory_embs) >= move_to_cpu_every:
+                   all_embs = torch.cat((all_embs, in_gpumemory_embs.cpu().detach()), dim=0)
+                   in_gpumemory_embs = torch.empty([0], device=self._speaker_model.device)

+             all_embs = torch.cat((all_embs, in_gpumemory_embs.cpu().detach()), dim=0)

            del test_batch

+       del in_gpumemory_embs

        with open(manifest_file, 'r', encoding='utf-8') as manifest:
            for i, line in enumerate(manifest.readlines()):
```

data transfert may not explain this amount of CPU usage, I keep investigating and will keep you udpated

**Comment by Thytu** [comment]
It appears there are two distinct issues at play:

**Hardware-Related Issue**: We've identified a hardware-related concern where the performance varies by a factor of three depending on the instance we use, despite employing the same code and software environment. We've initiated a support ticket with our cloud provider to investigate this unusual behavior.

**Issue with _extract_embeddings Function**: We've made some adjustments in the code, as indicated below:on. When applying the following changes we fix the 3x variation:

``` diff
+         move_to_cpu_every = 400
+         all_embs = torch.empty([0])
+         in_gpumemory_embs = torch.empty([0], device=self._speaker_model.device)

-       all_embs = torch.empty([0])

        for test_batch in tqdm(
            self._speaker_model.test_dataloader(),
            desc=f'[{scale_idx+1}/{num_scales}] extract embeddings',
            leave=True,
            disable=not self.verbose,
        ):

            test_batch = [x.to(self._speaker_model.device) for x in test_batch]
            audio_signal, audio_signal_len, labels, slices = test_batch

            with autocast():
                _, embs = self._speaker_model.forward(input_signal=audio_signal, input_signal_length=audio_signal_len)
                emb_shape = embs.shape[-1]
                embs = embs.view(-1, emb_shape)

-               all_embs = torch.cat((all_embs, embs.cpu().detach()), dim=0)

+               in_gpumemory_embs = torch.cat((in_gpumemory_embs, embs), dim=0)

+               if len(in_gpumemory_embs) >= move_to_cpu_every:
+                   all_embs = torch.cat((all_embs, in_gpumemory_embs.cpu().detach()), dim=0)
+                   in_gpumemory_embs = torch.empty([0], device=self._speaker_model.device)

+             all_embs = torch.cat((all_embs, in_gpumemory_embs.cpu().detach()), dim=0)

            del test_batch

+       del in_gpumemory_embs

         with open(manifest_file, 'r', encoding='utf-8') as manifest:
+
+            move_to_cpu_every = 400
+            on_gpu_embeddings = {}
+            all_embs = all_embs.to("cuda")
+
             for i, line in enumerate(manifest.readlines()):
                 line = line.strip()
                 dic = json.loads(line)
                 uniq_name = get_uniqname_from_filepath(dic['audio_filepath'])
-                if uniq_name in self.embeddings:
-                    self.embeddings[uniq_name] = torch.cat((self.embeddings[uniq_name], all_embs[i].view(1, -1)))
+
+                if uniq_name in on_gpu_embeddings:
+                    on_gpu_embeddings[uniq_name] = torch.cat((on_gpu_embeddings[uniq_name], all_embs[i].view(1, -1)))
                 else:
-                    self.embeddings[uniq_name] = all_embs[i].view(1, -1)
+                    on_gpu_embeddings[uniq_name] = all_embs[i].view(1, -1)
+
                 if uniq_name not in self.time_stamps:
                     self.time_stamps[uniq_name] = []
                 start = dic['offset']
                 end = start + dic['duration']
                 self.time_stamps[uniq_name].append([start, end])
 
+                if len(on_gpu_embeddings) >= move_to_cpu_every:
+                    for uniq_name in on_gpu_embeddings.keys():
+                        if uniq_name in self.embeddings:
+                            self.embeddings[uniq_name] = torch.cat((self.embeddings[uniq_name], on_gpu_embeddings[uniq_name].to("cpu")))
+                        else:
+                            self.embeddings[uniq_name] = on_gpu_embeddings[uniq_name].to("cpu")
+
+                    del on_gpu_embeddings[uniq_name]
+
+
+            for uniq_name in on_gpu_embeddings.keys():
+                if uniq_name in self.embeddings:
+                    self.embeddings[uniq_name] = torch.cat((self.embeddings[uniq_name], on_gpu_embeddings[uniq_name].to("cpu")))
+                else:
+                    self.embeddings[uniq_name] = on_gpu_embeddings[uniq_name].to("cpu")
+
+            del on_gpu_embeddings
+
         if self._speaker_params.save_embeddings:
```

It's intriguing to observe such a significant difference in performance simply by utilizing torch.cat on the GPU. We'd love to hear your thoughts on this matter and whether there are any insights or suggestions you can provide.

Your input and expertise are greatly appreciated as we work to address these challenges.

Best regards,
Valentin DM.

**Comment by Thytu** [comment]
Issue wasn't related to NeMo, we simply changed `torch.cat` to `np.cat` and it reduces our CPU consumption for some of our instances.

Closing the discussion.
